### ISSUE TITLE
Allow HTML & CSS file imports for components

### ISSUE THREAD

**Issue by bob2517** [Active-CSS/active-css#222]
As an alternative to html {} inside components, have a html() parameter in the component statement that imports html file in the html {} location without having to do manual ajax and {$STRING} rendering.

**Comment by dragontheory** [comment]
Will this handle CORS requests?

**Comment by bob2517** [comment]
Hmmm... good point. It will probably need additional options. Have you been running into any CORS issues, or are just thinking ahead?

It should be fine for same-server requests if there has not be any additional security added on the back-end. All the ajax in the core runs through the same code, so it should be fine if your ajax normally works and the HTML files are being stored on the same server.

Hang on a sec though. Are you running a server? Or did you want to test this on a local computer using file://?

**Comment by dragontheory** [comment]
> Have you been running into any CORS issues, or are just thinking ahead?

No problems. Just thinking ahead because I've had issues with that in the past. 

I am running on a local server but I am also demo'ing in CodePen, which doesn't allow file hosting. 

It would certainly be nice to be able to just drag/drop index.html into a browser and have everything just work out-of-the-box, but please don't hold up dinner for the desert! lol

**Comment by bob2517** [comment]
Ok - I'll get it working with ajax first and we'll try it out and see what it can and can't handle.

There is an alternative solution for offline websites with the command load-as-ajax, which I use for the offline version of the docs site to simulate ajax calls (apologies because that command is not very well-documented). You essentially put HTML into template tags and simulate an ajax request on those to "load" them. That could be tied into this new method. Will take a look at that once the ajax stuff is done.

**Comment by bob2517** [comment]
Note to self: immediately after this ticket is done, do the issue where it talks about using variables, etc. inside server-side HTML, as that will be probably the next thing that crops up regarding this method. Most of the components on the docs site utilise variables in the HTML...

**Comment by bob2517** [comment]
Also needs css option for pulling in CSS, and also the observe option from create-element to really get rid of unnecessary typing.

I'm not going to complicate this issue by going into too many specifics, but the final result for a basic component will look something like:

```
@component clock-component html("/examples/clock.html") {
}
```

That would pull in /examples/clock.html whenever <clock-component></clock-component> was found on the page.

It's more complicated that this, as there are lots of things you can do with components that I need to account for, like shadow DOMs, private scopes, variables, etc. but that's the basic starting code for this upgrade. Most of the code for this is written already - it's more a case of sorting out syntax rules and stuff like that and tying in the ajax. Once that is all done everything should just "magically" work.

**Comment by bob2517** [comment]
I'm going to leave this as my test case for now. I'll come back and integrate create-element into the component syntax when the html option is working, as I said I would get that done this weekend and it's an easier immediate target. That will include full event and scoping capability.

```
body:init {
	create-element: clock-component clockComponent;
}

@component clockComponent html("/examples/clock.html") {
...
```

**Comment by bob2517** [comment]
Note to self: it's probably going to need caching options like ajax too. For now, whenever the component is run, it will always fetch dynamically. But I may reverse this and put in a no-cache option as I think caching will probably be the most wanted default. There's no point going multiple times to the server for the same content, but it could be dynamically generated HTML, so the future no-cache option is indicated.

**Comment by bob2517** [comment]
Note to self: having a "template(#myHTML)" style of option will solve pulling in HTML from the page as an alternative for offline websites using components like this.

**Comment by bob2517** [comment]
Note to self: Add the @function issue to the 2.10.0 release. The example I'm using has a custom event that would be better off as an ACSS function call, just solely because the syntax currently looks a bit weird.

**Comment by bob2517** [comment]
Note to self: Nested components using this method will need to wait until the HTML loads, before it can find out what to do with the inner component, because there may be another one to render inside it. So this method is going to be slower than other methods because of the round-trips to the server necessary for the case of nested components. To mitigate this liability, I'm going to put in an option to load and cache inner HTML for components as part of config loading. That way, there shouldn't be a noticeable stagger when rendering components dynamically. It will, however, affect the time that it takes to load the initial config. I don't see a cleaner way of doing that without the internals getting overly complicated trying to deal with the inevitable race conditions of pre-loading HTML for components at the same time as trying to render it.
Having a "bundle HTML" type of command could accept one file containing multiple HTML components, but then that's a step towards complexity that doesn't make a lot of sense when you could just use the upcoming template() method to pre-render all the HTML you need in your main file, which would process quicker anyway.

**Comment by bob2517** [comment]
There will be a "pre-load" option in the @component syntax (not right now - later), which will preload the HTML file in a component at the point of config load, rather than at the point when the component is rendered. If pre-load is not specified, HTML will be loaded dynamically and cached. There will also be a "no-cache" option (not right now - later) which will force a fresh reload of HTML whenever the same component gets rendered.

**Comment by bob2517** [comment]
Note to self: Setting up options in the core, "preload-files" and "nocache-files" are stored but don't do anything yet. These could change. All other options are also set up ready to be implemented.

**Comment by dragontheory** [comment]
Was reading through your comments and it occurred to me that someone may want to call more than one HTML file or mix the different ways to get HTML.

Something like this?
```
htmlFile(/myfile.html)
html(selector containing HTML already rendered)

/* and inside the component itself if neither of those are used: */
html {
<p>Would this paragraphappearafterthe ajax html file?</p>
}
```

**Comment by bob2517** [comment]
Ok, point noted. Let's keep the syntax for this simple for now, and then see what is indicated in practice when using them. A future upgrade should be a bit more obvious once this basic syntax is sorted out.

**Comment by bob2517** [comment]
I'm re-jigging the syntax for html() itself to allow for the ajax options, like post, get, nocache, etc. I'm not far off getting this working now. Should have something by the end of the day hopefully.

**Comment by bob2517** [comment]
Note to self: Error handling for failing HTML load will need to be set up next weekend with a handling which has yet to be worked out.

**Comment by bob2517** [comment]
Got the first draft done, with the html parameter working (it seems).

Going to put onto the branch in a moment and see what happens next... famous last words.

Basic syntax is:

```
@component clockComponent html("/examples/clock.html" get) {
```

```
@component clockComponent html("/examples/clock.html" post) {
```

It allows all the parameters of the ajax command within the html option. So when I add more to the ajax command, it will reflect in the html option.

I've not done the css() option or any of the other options yet. Variable scoping and events will work though.

**Comment by bob2517** [comment]
Just tested with reactive variables in the HTML file. It even works with that. I disallow it for regular HTML loaded with ajax. I just blew my own mind :) Not sure whether I should leave that in or not - I need to think a bit more about any potential implications.

Retiring for the eve now. May resume during the week, otherwise 't will be next weekend for the rest.

**Comment by bob2517** [comment]
The test case is this below. There is still room for improvement - @function will replace the use of the custom event, and the create-element instruction will disappear when this issue is complete.

The html file contains:
```
<div>Time: {{clockTime}}</div>
```

And the currently working ACSS is:
```
body:init {
	create-element: clock-component clockComponent;
}

@component clockComponent html("/examples/clock-component.html" get) private {
	&:beforeComponentOpen {
		~clock { trigger: tick; }
	}
	~clock:tick {
		var: clockTime new Date().toLocaleTimeString() every 1s label clockTick;
	}
	div:click {
		@if has-class(self .clockOff) {
			remove-class: .clockOff;
			~clock { trigger: tick; }
		} @else {
			cancel-timer: clockTick;
			add-class: .clockOff;
		}
		add-class: .clockPress;
		remove-class: .clockPress after 100ms;
	}
}
```

which will become something like this leaner version:

```
@component clock-component html("/examples/clock-component.html" get) private {
	&:beforeComponentOpen {
		func: tick;
	}
	@function tick {
		var: clockTime new Date().toLocaleTimeString() every 1s label clockTick;
	}
	div:click {
		@if has-class(self .clockOff) {
			remove-class: .clockOff;
			~clock { trigger: tick; }
		} @else {
			cancel-timer: clockTick;
			add-class: .clockOff;
		}
		add-class: .clockPress;
		remove-class: .clockPress after 100ms;
	}
}
```

**Comment by bob2517** [comment]
Works with the shadow DOM option too. The CSS link tag can be in the HTML or it can be embedded. I'll still do the css() option, just in case the developer wants that in a separate file too from the HTML file. That way, you could have different CSS for the same component. Separation of concerns. Dragon theory possibly gives the thumbs up for that.

**Comment by dragontheory** [comment]
You know it bro!

**Comment by bob2517** [comment]
Next, implement the css() option. Both html and css files will need to be loaded before the component is rendered, so we don't get unwanted flicker of CSS when a component is drawn.

Shadow DOMs can actually have their own link tags in the HTML, which is a different technique. It would potentially be slower to render this way though, as there would be a definite stagger in calling the server. Only when the HTML was rendered would the CSS be drawn, which personally I think is a bit of a raw prawn.

I'll get css() working with regular non-shadow DOM components first and see where it leads with regards any additional shadow DOM technique.

**Comment by bob2517** [comment]
Well, that was easy to set up. Works with shadow DOM components too - no flicker. It waits until both files have been retrieved before the component is drawn.

Will commit this in a sec and then move onto getting it working with HTML templates, which will provide a faster render time by using HTML that is already rendered on the page inside standard template tags.

This is a bit of a game changing one, this one about importing HTML and CSS. It avoids getting into messy HTML and CSS coding inside ACSS config, which is something I've never been particularly happy about. Thank you Dragon Theory for requesting it.

**Comment by bob2517** [comment]
The css() option is now on the branch. It doesn't have to be used, but it's there anyway. It's a good solution if it's a shadow DOM component and you want to keep your CSS for it in a separate file.

**Comment by bob2517** [comment]
I'm splitting the use of pre-rendered template tags to store content, to facilitate separated pre-rendering when using the technique of compiling pages on the server.

Where ajax uses html(filename) and css(filename) options, templates will have corresponding html-template(template tag selector) and css-template(template tag selector) options.

These template options will be able to work alongside html() and css() options for maximum flexibility, but they won't be rendering together in any particular order. We are simply wanting to output HTML content together with CSS content (which gets put into a style tag automatically by the core). So the recommended combinations will be:

html() + css()
html-template() + css-template()
html() + css-template()
html-template() + css()

You could, of course, just put the CSS into a style tag directly into any imported HTML, and not bother with the css() or css-template() options.

By default, ACSS variables will not be allowed in these imported files. This allows for safer rendering from database content on the server. Eg. a user may submit a blog post containing some ACSS variables. We don't want these evaluating by default if then later rendered in code from a developer who forgot to put the unplanned-for option "disallow-vars" into their ACSS component. It should be the other way around. The developer should specifically allow the use of variables in components, so to allow variables, allow-vars needs to be an option in the component, which puts the control in the hands of the developer by enforcing a conscious decision to allow ACSS variables to be recognised as such in their components.

**Comment by dragontheory** [comment]
> Thank you Dragon Theory for requesting it.

Many will benefit from your time and effort. Thank YOU for building it.

**Comment by bob2517** [comment]
No... thank YOU! Lol :)

**Comment by bob2517** [comment]
Note to self:

Put a warning in the docs. When using ACSS config in an iframe (like I'm doing for my tests) and the HTML or CSS is stored in templates outside of the iframe, the parent selector needs to be employed in the html-template option. Eg. "parent -> #myHTMLTemplate". It took me a while to work out what was going on... Now that I know what's going on, I'm just going to put the template into the iframe itself, where the action is happening.

**Comment by bob2517** [comment]
Also note in the docs for this, that CSS that comes from a template should have a style tag in the template, otherwise it isn't proper valid HTML.

**Comment by bob2517** [comment]
html-template() and css-template() are now done, and I'm about to do a commit for those.

I've set up up now so that if there is additionally an html { } inside the component itself, then that will render above any other html or css that is loaded via component options.

Time will tell what implications that has... I'm leaving that there for now to see what happens. Might be useful, I dunno.

**Comment by bob2517** [comment]
html-template() and css-template() options are now on the branch.
These would look like this in the HTML:
```
<clock-component></clock-component>

<template id="clockTemplateHTML">
<div>Time: {{clockTime}}</div>
</template>

<template id="clockTemplateCSS">
	<style>
		div {
			background-color: #699869;
			color: #f3fff3;
			width: 340px;
			height: 100px;
			margin: 30px 0 40px 20px;
			padding-top: 44px;
			font-size: 36px;
			text-align: center;
			border-bottom: 4px solid #2f662f;
			border-radius: 50%;
			transform: skewX(15deg);
			box-shadow: -13px 8px 12px 1px rgba(0,0,0,0.16), 0 2px 10px 0 rgba(0,0,0,0.12);
			cursor: pointer;
			user-select: none;
			-webkit-tap-highlight-color: transparent;
			transition: border-width 100ms linear, transform 100ms linear, box-shadow 100ms linear;
		}

		div.clockPress {
			background-color: #74a074;
			color: #e7f7e7;
			transform: skewX(12deg);
			box-shadow: -4px 5px 12px 0 rgba(0,0,0,0.16), 0 2px 10px 0 rgba(0,0,0,0.12);
			border-width: 1px;
		}

		div.clockOff {
			background-color: #a2a2a2;
			border-color: #4d4d4d;
			color: #bfbfbf;
		}
	</style>
</template>

```

With the component looking like - as it stands currently before further work on this:
``
body:init {
	create-element: clock-component clockComponent;
}

@component clockComponent
		html-template(#clockTemplateHTML)
		css-template(#clockTemplateCSS)
	private {
	&:beforeComponentOpen {
		~clock { trigger: tick; }
	}
...etc
```

Next will be setting it up so that the create-element command is not needed. I need to do some household chores now which I've been putting off all weekend, and will hopefully come back to this later on.

**Comment by bob2517** [comment]
Note to self: Re implying create-element within component declaration: just call create-element during a pre-user init event if the component name is at least double-hyphenated. If it's double-hypenated, assume that the double-hypenated component name is also the tag name. Just make it clear in the docs that component names, if they are being used in a non-custom element way, should be camel-cased with no hypen. Component naming convention has been implied by the ACSS docs, but not officially stated. Custom elements should always be double-hypenated per the HTML spec, and that definitely is mentioned in the docs. Internally, when rendering the component, it needs a name, and I think the component name can be the same as the custom element tag name - it should be fine with hyphens - may need tweaking if not currently.

**Comment by bob2517** [comment]
I'm going to complete this issue today, if it's the last thing that I do!

**Comment by bob2517** [comment]
Next to tackle - official creation of the custom element as part of the component syntax.

So instead of a component looking like:
```
body:init {
  create-element:  clock-component clockComponent;
}

@component clock-component
    html-template(#clockTemplateHTML)
    css-template(#clockTemplateCSS)
    private
  {
  &:beforeComponentOpen {
    ~clock { trigger: tick; }
  }
...etc
}
```
It will now allow a more direct syntax:
```
@component clock-component
    html-template(#clockTemplateHTML)
    css-template(#clockTemplateCSS)
    private
  {
  &:beforeComponentOpen {
    ~clock { trigger: tick; }
  }
...etc
}
```

I'll expand the component reference syntax to allow for {|clock-component}. That way we shouldn't need the camelCase reference at all. But that will still be allowed for backward compatibility and also for non-custom element components.

After that, the observe() option from create-element for observing attribute changes as per the native web component rules will be placed directly into the @component statement, so that the correct setups are done behind the scenes.

So in summary:
1) Get the element working as a name in @component.
2) Get the component reference variable "{|component}" allowing a dash.
3) Add observe() into the @component statement and tie it into create-element.
4) Tie in create-element automatically so everything gets set up.

**Comment by dragontheory** [comment]
> After that, the observe() option from create-element for observing attribute changes as per the native web component rules will be placed directly into the @component statement, so that the correct setups are done behind the scenes.

Ensuring syntax accuracy via automation while not obfuscating contextual meaning. I like it!

**Comment by dragontheory** [comment]
> If the component name is double-hypenated, assume that the double-hyphenated component name is also the tag name

This would be nice! 

> Component naming convention has been implied by the ACSS docs, but not officially stated.

I did what most DEVs do I think. Just copied/pasted the pattern you had in your documentation or examples you've given me. I think that is a strong convention.

**Comment by dragontheory** [comment]
> Next to tackle - official creation of the custom element as part of the component syntax.

So, in my example, this would eliminate the need for my components.acss that ties hyphenated HTML elements to ACSS equivalents? The association would be implied?  
```
body:init {
  create-element:
    list-app listApp,
    list-dataset listDataset,
    list-details listDetails,
    list-pagination listPagination,
    list-results listResults,
    page-container pageContainer,
    page-header pageHeader,
    panel-nav panelNav,
    panel-intro panelIntro,
    panel-main panelMain,
    panel-splitter panelSplitter,
    panel-aside panelAside,
    panel-aside2 panelAside2,
    page-footer pageFooter;
}
```

**Comment by bob2517** [comment]
Yep. You will just need to rename your components to match the tag names and then you won't need this file at all.

**Comment by dragontheory** [comment]
> I think that could take the event-driven method to the next level though - checking both the state of the DOM and anything else that can be set up in an @if statement, in just one @observe event statement.
>
> Need some feedback. DT?

Timely that you are addressing this...

This is getting down into the weeds a bit but that is where we need to go if we are to make sure all this works with real world scenarios.

In my example, I have styles and layouts pre-setup when key classes or combination of classes are in the `<body>`. Those classes are added to or removed from `<body>` when certain DOM content changes, via ACSS `@observe`. 

That's cool and all, but the piranha in the water here (to borrow a metaphor) is that not ALL changes in the DOM layout are due to content changes.  

For example, the class `.aside2` is added to `<body>` when the `<user-name>` DOM element is clicked, allowing `<panel-aside2>` (user settings panel) to be visible. 

While the other layout classes can only be added/removed conditionally and in workflow (left to right) sequence, the `.aside2` class can be added/removed from `<body>` at any time. 

As the number of different layouts and combination of layouts grow, I forsee this logic both on the CSS and the ACDS side [getting out of hand](https://i.imgflip.com/5yudos.png) ... _so to speak_.

So, this will be really helpful me thinks.

**Comment by bob2517** [comment]
Lol! meme.

Great feedback, thank you. That's good enough for me - I'm implement it and see what happens.

**Comment by bob2517** [comment]
The create-element command is no longer needed now (for components) on the branch.

For example, a component can just be written like this now. HTML:
```
<basic-clock></basic-clock>
```

with ACSS:
```
@component basic-clock private {
	&:beforeComponentOpen {
		var: clockTime new Date().toLocaleTimeString() every 1s;
	}
	html {
		<div>Time {{clockTime}}</div>
	}
}
```

To make this possible, the observe() option from the create-element command (for monitoring attribute changes in line with the W3C spec) is now accepted directly in the @component statement.

The create-element command is still useful if you want to officially create elements and are not using components, so backward-compatibility is maintained and create-element will not be deprecated.

I also fixed an issue with the parent selector when used in combination with a complex selector. So "host < div p" will find the component host element, then find the closest div, then go back down the DOM looking for the remainder following the CSS combinator rules. The following are also now supported to give the full predictable range of combinator support, "host < div >p", "host < div + p" and "host < div ~ p". This is now on the branch.

**Comment by bob2517** [comment]
The only thing left on this issue is to add "accept-vars" as an option for the component, and remove the current default behaviour of allowing ACSS vars. ACSS vars will only be automatically rendered as such if accept-vars is an option in the component. This helps prevent variable injection from user input rendered from a database on the server.

This does not affect backward compatibility for any release prior to 2.10.0, but it does change behaviour on the branch, so the "accept-vars" option will need to be added for 2.10.0 components containing variables in external files or templates if this new branch feature has been employed.

**Comment by dragontheory** [comment]
> The create-element command is no longer needed now (for components) on the branch.

To confirm, I no longer need `components.acss` because all it has is this code in it? 
```
body:init {
  create-element:
    list-app listApp,
    list-dataset listDataset,
    list-details listDetails,
    list-graphics listGraphics,
    list-layouts listLayouts,
    list-pagination listPagination,
    list-results listResults,
    list-themes listThemes,
    page-container pageContainer,
    page-header pageHeader,
    page-search pageSearch,
    panel-nav panelNav,
    panel-intro panelIntro,
    panel-main panelMain,
    panel-main-count panelMainCount,
    panel-splitter panelSplitter,
    panel-aside panelAside,
    panel-aside2 panelAside2,
    page-footer pageFooter;
}
```

**Comment by dragontheory** [comment]
> For example, a shadow DOM clock component can just be written like this now. HTML:
> 
> ```
> <basic-clock></basic-clock>
> ```

Beautiful in it's simplicity.

**Comment by bob2517** [comment]
> > The create-element command is no longer needed now (for components) on the branch.
> 
> To confirm, I no longer need `components.acss` because all it has is this code in it?
> 
> ```
> body:init {
>   create-element:
>     list-app listApp,
>     list-dataset listDataset,
>     list-details listDetails,
>     list-graphics listGraphics,
>     list-layouts listLayouts,
>     list-pagination listPagination,
>     list-results listResults,
>     list-themes listThemes,
>     page-container pageContainer,
>     page-header pageHeader,
>     page-search pageSearch,
>     panel-nav panelNav,
>     panel-intro panelIntro,
>     panel-main panelMain,
>     panel-main-count panelMainCount,
>     panel-splitter panelSplitter,
>     panel-aside panelAside,
>     panel-aside2 panelAside2,
>     page-footer pageFooter;
> }
> ```

Yes, that's correct. You can remove those. All you then need to do is make sure the component names match the element, so for example:

```
@component pageFooter {
```

becomes:
```
@component page-footer {
```

And then do that for all your components.

I agree, it does look a lot neater. It took a while to get that syntax right...

**Comment by bob2517** [comment]
Almost there for a release soon, then I'll tackle your page splitter. The target for the 2.10.0 release with the existing issues finished is 10th April.

**Comment by dragontheory** [comment]
> Yes, that's correct. You can remove those. All you then need to do is make sure the component names match the element,

Made the changes and they are working great!

**Comment by bob2517** [comment]
Cookies are not stored in directories - the directory is more like a scope for the cookie - the cookie itself is kept in memory. You can inspect the values of cookies from DevTools (in Chrome - under the Application tab, in the Storage section).

Try setting a cookie in there - set path to ('/') to make sure that your whole website can get to it. Then have a look in DevTools and see if it's there. You might need to refresh the page.

What do you want to then do with the cookie, out of interest? There are some conditionals in ACSS for checking for cookies, but there's probably a case for adding more cookie functionality.

**Comment by bob2517** [comment]
> > Yes, that's correct. You can remove those. All you then need to do is make sure the component names match the element,
> 
> Made the changes and they are working great! Thank you!

That's great! :)

**Comment by dragontheory** [comment]
> Cookies are not stored in directories - the directory path option is more like a scope for the cookie - the cookie itself is kept in memory. You can inspect the values of cookies from DevTools (in Chrome - under the Application tab, in the Storage section).
> 
> Try setting a cookie in there - set path to ('/') to make sure that your whole website can get to it. Then have a look in DevTools and see if it's there. You might need to refresh the page.

Ah OK. 

Going off of your example: 
```
set-cookie: name(cookieNoticeSeen) value("y") expires("Year") path("/") sameSite("Strict") secureIfHttps;
```
I now have a settings panel that changes the themes like you showed me. I want to set a cookie so I don't have to keep changing it. lol

**Comment by bob2517** [comment]
Apologies - I gotta rush out to a class, but check out the local/session storage solution that I mentioned earlier - I would use that instead of a cookie. Will be back in a couple of hours. If you definitely want to go with a cookie, have a read up on cookies - the ACSS implementation just copies the standard way of doing it.

**Comment by dragontheory** [comment]
Also, why would I want to remove a cookie?

**Comment by dragontheory** [comment]
> check out the local/session storage solution that I mentioned earlier - I would use that instead of a cookie

Ah OK. Thanks!

**Comment by bob2517** [comment]
> Also, why would I want to remove a cookie?

You might store several preferences in a cookie and want to remove one. Or the user wants to turn advertiser tracking cookies off, so you need to remove them. The option is there - it depends on the use case.

But yeah, with the example on this page:
https://activecss.org/manual/if-cookie-equals.html

If you were to go down that route, you could do something like this:
```
/* Set the cookie if it isn't there already */
body:not-if-cookie-exists(theme):init {
  set-cookie: name(theme) value("light") expires("Year") path("/") sameSite("Strict") secureIfHttps;
}

/* Set up the CSS variables depending on the theme from the cookie. */
body:init {
  :root {
    @if cookie-equals(theme "light") {
      --myBackground: #fff;
    } @else if cookie-equals(theme "dark") {
      --myBackground: #111;
    }
  }
}

/* Set up the theme chooser. */
.themeChooser:click {
  set-cookie: name(theme) value("{@data-theme}") expires("Year") path("/") sameSite("Strict") secureIfHttps;
}
```
```
<button class="themeChooser" data-theme="light">light</button>
<button class="themeChooser" data-theme="dark">dark</button>
```

You could follow the same approach with the var command using local or session storage. The logic is the same, but the commands would be different.

**Comment by bob2517** [comment]
This is the CSS variables demonstration page:
https://activecss.org/manual/css-variables-example.html

**Comment by bob2517** [comment]
> So is "Year" filler for a year or do I have to set a date with a year?

From the docs:
```
expires: This is how long the cookie is going to last for. Possible values are "Infinity", "Year", "Month", "Hour", "Second", a date string in Coordinated Universal Time (UTC) format, or friendlier things like "3 months", "-1 hour", "2 years", "1 year", "+60 seconds", "45 minutes". If you need a more custom solution for a date parameter don't forget you can use the "{= JavaScript date expression =}", which you can use in all action commands, eg. 'expires("{= new Date().toUTCString() =}")'.
```

So to answer your question, yes you just put "Year". Or you could put "2 months" or "2 years".

**Comment by bob2517** [comment]
Reviewing this issue - a few things left to wrap up tomorrow to close the issue:

1) accept-vars as a means of allowing ACSS vars in components. This is currently what happens with the new functionality on the branch, but it needs to only happen when accept-vars is set as an option, for database-injection security reasons. The developer needs to be in control of automatic variable declaration and take precautions on the server for escaping ACSS variables when rendering user submitted text.
2) disallow ACSS vars from imported html in components as default behaviour. That should just be a checking step after finishing point 1.
3) accept-vars as a means of allowing ACSS vars in ajax HTML responses.
4) disallow ACSS vars from ajax HTML responses as default behaviour. The core should already be doing that, so this should just be a checking step too.
5) Another one mentioned earlier in the issue by myself is to allow {|basic-clock} to render the inner contents of a component that is already pre-defined. I'm not convinced at this point though if this is even needed, and I don't have a use case, so I may skip that one for now. It'll be easy to add it in later if needed.

**Comment by bob2517** [comment]
This is almost done - should have a commit by the end of the day. Then tomorrow I'll work on the documentation and try and get that wrapped up so we can get the new release out.

**Comment by bob2517** [comment]
Accept-vars will also work with JSON returns, so variables can be allowed to be substituted if the JSON value contains things like {myValue}. Should work with different variable primitives. That's the plan, anyway.

**Comment by bob2517** [comment]
I think this accept-vars stuff is wrapped up now. I just need to set up official tests to make sure I've covered all scenarios.

Alternative syntaxes:

For global component allowing of ACSS vars in imported files (accept-vars):
```
@component clock-component-with-vars
		html(/examples/clock-component-with-vars.html get)
		css(/examples/clock-component-with-vars.css get)
	private shadow accept-vars {
```

Specific vars allowed from certain files can be limited to the request, like here for example which allows them on the CSS file only and not the HTML file:
```
@component clock-component-with-vars
		html(/examples/clock-component-with-vars.html get)
		css(/examples/clock-component-with-vars.css get accept-vars)
	private shadow {
```

For regular ajax requests:
```
ajax: "/examples/clock-component-with-vars.html" get html accept-vars;
ajax: "/examples/clock-component-vars" get json accept-vars;
```

There was an error where the {$STRING} html variable wasn't getting declared as expected with the await option, which I've fixed as part of this issue.

There is quite a lot of documentation upgrading to do on components, so I'm putting the branch release date back to next weekend and we'll see how it goes with that. This was a bit fiddlier to implement than I suspected thanks to component scoping issues. Should be all done now though. This accept-vars stuff won't be put on the branch until I've done the docs with comprehensive examples and am certain that it all works.

There's a possibility that it can allow multiple html and css files. Probably will do that after this release though, otherwise it's just going to drag out.

**Comment by bob2517** [comment]
I'm going to be putting this onto the branch by the end of the day. When it does go on the branch, @dragontheory, you are going to need to use the accept-vars option on any new @component statement (or any ajax command) that imports code which tries to render ACSS variables. Look at the example above for how to do that.

So if you upgrade to the latest branch (when I update it) and you see something breaking on working components, then that is possibly the reason. Just attach the accept-vars option to the ajax request in the @component statement.

**Comment by bob2517** [comment]
Now fully on the latest 2.10.0 branch.

**Comment by dragontheory** [comment]
> @dragontheory, you are going to need to use the accept-vars option on any new https://github.com/component statement (or any ajax command) that imports code which tries to render ACSS variables from within the code.

Very good. Will make the change. Thank you for the heads-up!
